File: dart/flowdux/lib/src/flow_holder_middleware.dart
import 'dart:async';

import 'action.dart';
import 'middleware.dart';
import 'store_logger.dart';
import 'strategy/execution_strategy.dart';

/// Internal middleware that processes FlowHolderAction with its execution strategy.
///
/// This middleware is automatically added to the end of the middleware chain
/// by the Store. It handles:
/// - Applying the action's [ExecutionStrategy]
/// - Recursive processing of nested FlowHolderActions
/// - Logging of emitted actions
///
/// The [dispatch] function is used when a FlowHolderAction has
/// [FlowActionDelivery.dispatch] delivery mode.
class FlowHolderMiddleware<S, A extends Action> extends Middleware<S, A> {
  final StoreLogger<S, A> _logger;
  final void Function(A) _dispatch;

  /// Cached wrapped processors for FlowHolderActions, keyed by runtimeType.
  final Map<Type, Stream<A> Function(S, A)> _wrappedProcessors = {};

  /// Creates a [FlowHolderMiddleware] with the specified [logger] and [dispatch].
  FlowHolderMiddleware(StoreLogger<S, A> logger, void Function(A) dispatch)
      : _logger = logger,
        _dispatch = dispatch;

  @override
  Stream<A> process(S Function() getState, A action) {
    if (action is! FlowHolderAction) {
      return Stream.value(action);
    }

    final wrapped = _getOrCreateWrappedProcessor(action);

    switch (action.delivery) {
      case FlowActionDelivery.emit:
        return wrapped(getState(), action).asyncExpand(
          (innerAction) => _processRecursively(getState, innerAction),
        );

      case FlowActionDelivery.dispatch:
        // Re-dispatch through full pipeline; nested FlowHolderActions
        // will be processed when they reach this middleware again
        return wrapped(getState(), action).asyncExpand((innerAction) {
          _dispatch(innerAction);
          return const Stream.empty();
        });
    }
  }

  /// Recursively processes nested FlowHolderActions.
  Stream<A> _processRecursively(S Function() getState, A action) {
    _logger.onFlowHolderActionEmitted(action);
    if (action is FlowHolderAction) {
      return process(getState, action);
    }
    return Stream.value(action);
  }

  /// Gets or creates a wrapped processor for the given FlowHolderAction type.
  Stream<A> Function(S, A) _getOrCreateWrappedProcessor(FlowHolderAction action) {
    return _wrappedProcessors.putIfAbsent(action.runtimeType, () {
      Stream<A> baseProcessor(S state, A a) {
        return (a as FlowHolderAction).toStreamAction().cast<A>();
      }
      return action.strategy.wrap<S, A, A>(baseProcessor);
    });
  }
}

File: kotlin/flowdux/src/commonMain/kotlin/io/flowdux/FlowHolderMiddleware.kt
package io.flowdux

import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.FlowCollector
import kotlinx.coroutines.flow.emitAll
import kotlinx.coroutines.flow.flow
import kotlinx.coroutines.flow.flowOf
import kotlinx.coroutines.flow.map
import kotlinx.coroutines.flow.onEach
import kotlinx.coroutines.flow.transform
import kotlin.reflect.KClass

/**
 * Internal middleware that handles [FlowHolderAction] processing.
 *
 * This middleware is automatically added as the last middleware in the Store's
 * middleware chain. It expands FlowHolderActions into their streams of actions
 * and applies the action's [ExecutionStrategy].
 *
 * For nested FlowHolderActions (a FlowHolderAction that emits another FlowHolderAction),
 * processing is done recursively within this middleware, not re-dispatched through
 * the entire middleware chain.
 *
 * @param dispatch function to re-dispatch actions through the full Store pipeline.
 *   Used when a FlowHolderAction has [FlowActionDelivery.Dispatch] delivery mode.
 */
internal class FlowHolderMiddleware<S : State, A : Action>(
    private val logger: StoreLogger<S, A>,
    private val dispatch: (A) -> Unit,
) : Middleware<S, A> {

    override val name: String = "FlowHolderMiddleware"
    override val processors: ActionProcessorMap<S, A> = emptyMap()

    /**
     * Cache of wrapped processors for each FlowHolderAction type.
     * The strategy wrapping is cached per KClass to maintain strategy state
     * (e.g., for TakeLatest, the same strategy instance tracks the current job).
     */
    private val wrappedProcessors = mutableMapOf<KClass<*>, WrappedProcessor<S, A>>()

    override fun process(getState: () -> S, action: A): Flow<A> {
        if (action !is FlowHolderAction) {
            return flowOf(action)
        }

        val wrapped = getOrCreateWrappedProcessor(action)

        return when (action.delivery) {
            FlowActionDelivery.Emit -> flow {
                wrapped.invoke(this, getState(), action)
            }.transform { innerAction ->
                if (innerAction is FlowHolderAction) {
                    // Recursive processing for nested FlowHolderActions
                    emitAll(process(getState, innerAction as A))
                } else {
                    emit(innerAction)
                }
            }

            FlowActionDelivery.Dispatch -> flow<A> {
                wrapped.invoke(this, getState(), action)
            }.onEach { innerAction ->
                // Re-dispatch through full pipeline; nested FlowHolderActions
                // will be processed when they reach this middleware again
                dispatch(innerAction)
            }.transform {
                // Don't emit anything; all inner actions are re-dispatched
            }
        }
    }

    @Suppress("UNCHECKED_CAST")
    private fun getOrCreateWrappedProcessor(action: FlowHolderAction): WrappedProcessor<S, A> {
        return wrappedProcessors.getOrPut(action::class) {
            val baseProcessor: suspend FlowCollector<A>.(S, A) -> Unit = { _, a ->
                emitAll(
                    (a as FlowHolderAction).toFlowAction()
                        .onEach { logger.onFlowHolderActionEmitted(it as A) }
                        .map { it as A }
                )
            }
            action.strategy.wrap(baseProcessor)
        }
    }
}

private typealias WrappedProcessor<S, A> = suspend FlowCollector<A>.(S, A) -> Unit
